File: backend/schemas/rating.py
"""Response shapes for `GET /rating/*`.

Both endpoints answer the same question from two sides — *is a learned aesthetic
head worth building* — and both are shaped so a client cannot render the headline
number without the caveat that qualifies it. Every rate is nullable and arrives
beside the counts it came from; below a floor the page shows the counts and no
percentage at all, because a ceiling computed from three pairs is noise wearing a
number.
"""
from pydantic import BaseModel


class SelfAgreementOut(BaseModel):
    """How often the same image got the same rating twice.

    Read `backend/ml/rating_metrics.py::self_agreement` for what each field
    counts. The short version: `pairs` are **consecutive** events per image, not
    first-versus-last (which scores 1 → 4 → 1 as perfect agreement);
    `singleton_*` restricts to pairs where **both** writes touched one image,
    which is the only subset the bulk-sweep bias cannot inflate; `cleared_pairs`
    are excluded from `pairs` entirely, because withdrawing a judgement is not a
    second opinion.
    """

    images_with_repeats: int
    pairs: int
    agreements: int
    rate: float | None
    singleton_pairs: int
    singleton_agreements: int
    singleton_rate: float | None
    bulk_pairs: int
    cleared_pairs: int
    # |Δ| = 1 versus |Δ| ≥ 2. A boundary wobble is not a Keep↔Cut flip.
    adjacent: int
    distant: int
    rate_within_1: float | None
    first_last_pairs: int
    first_last_agreements: int
    first_last_rate: float | None


class RatingEventsOut(BaseModel):
    total: int
    images_with_events: int
    images_with_repeats: int


class RatingSummaryOut(BaseModel):
    total: int
    rated: int
    unrated: int
    # Rated images whose pixels were rewritten since the rating was given, so the
    # judgement is about pixels that no longer exist.
    rating_stale: int
    # Keyed "1".."4" — JSON object keys are strings.
    by_rating: dict[str, int]
    events: RatingEventsOut
    self_agreement: SelfAgreementOut


class BoundaryAgreementOut(BaseModel):
    """One adjacent-tier boundary, e.g. Probably-not vs Probably.

    `auc` is P(a randomly drawn image from the upper tier scores above one from
    the lower), ties counted as 0.5. 0.5 is a coin flip, which is why the page
    draws its bar centred there rather than from zero.
    """

    boundary: str
    n_lo: int
    n_hi: int
    auc: float | None


class ScorerModelAgreementOut(BaseModel):
    # Nullable because nothing *enforces* `aesthetic_score IS NOT NULL ⟺
    # aesthetic_model IS NOT NULL` — the migration's backfill established it, and
    # any writer that sets a score without the marker breaks it again. A `None`
    # here means "scored by a writer that skipped the marker", **not** a model
    # named "unknown": the bucket is real and its `n` is counted, so
    # `sum(m.n) == scored_and_rated` holds. Skipping the row instead would break
    # that identity silently.
    model: str | None
    n: int
    spearman: float | None
    # The largest ρ *any* scorer could reach against this tier distribution. With
    # a four-level target the tie structure caps ρ below 1.0, so a bare 0.31 reads
    # as failure forever; "0.31 of a possible 0.97" is the honest form.
    spearman_ceiling: float | None
    # The most legible thing on the page: four flat tier means mean the scorer
    # knows nothing about your taste, and anyone can read four numbers.
    mean_by_rating: dict[str, float | None]
    # The count behind each of those means, on the wire for the same reason
    # `n_lo`/`n_hi` are: a mean over one image is indistinguishable from a mean
    # over two hundred, and it is not derivable from `boundaries` except by an
    # index trick that breaks the moment a boundary changes.
    n_by_rating: dict[str, int]
    boundaries: list[BoundaryAgreementOut]


class ScorerAgreementOut(BaseModel):
    rated: int
    scored_and_rated: int
    rated_unscored: int
    # A **list ordered by `n` desc**, not a dict keyed by marker, so a future
    # `head:{uuid}` producer lands in it without a schema change — matching what
    # `aestheticModelLabel` already says about unknown markers.
    models: list[ScorerModelAgreementOut]

File: frontend/src/api/rating.ts
import client from "./client";

/** How often the same image got the same rating twice.
 *
 *  **Counts, never a bare rate** — every `*_rate` is nullable and arrives beside
 *  the `pairs`/`agreements` it came from, because the day-one figure is a
 *  diagnostic and not a measurement. Three biases qualify it, each pushing a
 *  known way: *selection* (you re-rate what you disagree with) pushes it **down**;
 *  *anchoring* (the second look sees the previous answer) pushes it **up**;
 *  *bulk sweep* (select-all then press 1 writes events for images nobody looked
 *  at) pushes it **up** hardest, and `singleton_*` is what isolates it. */
export interface SelfAgreement {
  /** Images with two or more events — the whole universe of this statistic. */
  images_with_repeats: number;
  /** **Consecutive** event pairs, not first-versus-last: 1 → 4 → 1 is someone
   *  who cannot decide, and first-versus-last scores it as perfect agreement. */
  pairs: number;
  agreements: number;
  rate: number | null;
  /** Pairs where **both** writes touched exactly one image. The honest subset:
   *  a sweep cannot inflate it. */
  singleton_pairs: number;
  singleton_agreements: number;
  singleton_rate: number | null;
  /** `pairs − singleton_pairs`, stated rather than left to be inferred. */
  bulk_pairs: number;
  /** Pairs where either side cleared the rating. **Excluded** from `pairs`:
   *  withdrawing a judgement is not a second opinion. */
  cleared_pairs: number;
  /** |Δ| = 1 — a boundary wobble, not a Keep↔Cut flip. */
  adjacent: number;
  /** |Δ| ≥ 2. */
  distant: number;
  rate_within_1: number | null;
  first_last_pairs: number;
  first_last_agreements: number;
  first_last_rate: number | null;
}

export interface RatingSummary {
  total: number;
  rated: number;
  unrated: number;
  /** Rated images whose pixels were rewritten since — the judgement is about
   *  pixels that no longer exist. */
  rating_stale: number;
  /** Keyed "1".."4"; every tier present even at zero. */
  by_rating: Record<string, number>;
  events: { total: number; images_with_events: number; images_with_repeats: number };
  self_agreement: SelfAgreement;
}

export interface BoundaryAgreement {
  /** "1v2" | "2v3" | "3v4". */
  boundary: string;
  n_lo: number;
  n_hi: number;
  /** P(an image from the upper tier scores above one from the lower), ties at
   *  0.5. **0.5 is a coin flip**, which is why its bar is drawn centred there
   *  rather than from zero. Null when either tier is empty. */
  auc: number | null;
}

export interface ScorerModelAgreement {
  /** The `aesthetic_model` marker: "laion" | "v2_5" | a future `head:{uuid}`. */
  model: string;
  n: number;
  spearman: number | null;
  /** The largest ρ **any** scorer could reach against this tier distribution —
   *  with a four-level target the tie structure caps it below 1.0. Shown beside
   *  `spearman` because a bare 0.31 reads as failure forever, while "0.31 of a
   *  possible 0.97" is the truth. */
  spearman_ceiling: number | null;
  /** Keyed "1".."4"; null for a tier with nobody in it (not 0, which would read
   *  as "the scorer rates them at zero"). Four flat means is the most legible
   *  possible statement that a scorer knows nothing about your taste. */
  mean_by_rating: Record<string, number | null>;
  boundaries: BoundaryAgreement[];
}

export interface ScorerAgreement {
  rated: number;
  scored_and_rated: number;
  rated_unscored: number;
  /** Ordered by `n` desc. A **list**, not a dict, so a future `head:{uuid}`
   *  producer lands in it without a shape change. */
  models: ScorerModelAgreement[];
}

/** Both endpoints pool across every dataset and take no `dataset_id`: a head
 *  trained from pooled labels cannot live under one dataset. */
export const ratingApi = {
  summary: () =>
    client.get<RatingSummary>("/rating/summary").then((r) => r.data),
  scorerAgreement: () =>
    client.get<ScorerAgreement>("/rating/scorer-agreement").then((r) => r.data),
};
